android/src/main/java/com/tailscale/ipn/MainActivity.kt

```diff
@@ -182,12 +186,38 @@ class MainActivity : ComponentActivity() {
     }
   }
 
-  private fun login(url: String) {
-    // (jonathan) TODO: This is functional, but the navigation doesn't quite work
-    // as expected.  There's probably a better built in way to do this.  This will
-    // unblock in dev for the time being though.
-    val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
-    startActivity(browserIntent)
+  private fun login(urlString: String) {
+    lifecycleScope.launch {
+      try {
+        Notifier.state.collect { state ->
+          if (state > Ipn.State.NeedsMachineAuth) {
+            val intent =
+                Intent(applicationContext, MainActivity::class.java).apply {
+                  addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
+                  action = Intent.ACTION_MAIN
+                  addCategory(Intent.CATEGORY_LAUNCHER)
+                }
+            startActivity(intent)
```

Comment:
I'm pretty sure we need to stop the coroutine here, otherwise this will keep running even after we've launched the activity again? Or maybe the use of `lifecycleScope` above eliminated that need?

Reply:
I tested this and this coroutine does indeed keep running.

Reply:
aha. good catch, I thought lifecycleScope took care of it too